lib/ABIAnalyses/DeadReturnValuesOfFunctionCall.cpp

```diff
@@ -0,0 +1,142 @@
+#include <cstdio>
+#include <iostream>
+#include <llvm/IR/Instructions.h>
+#include <llvm/Support/Casting.h>
+
+#include "revng/ABIAnalyses/DeadReturnValuesOfFunctionCall.h"
+#include "revng/MFP/MFP.h"
+#include "revng/Support/revng.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/Instruction.h"
+
+namespace DeadReturnValuesOfFunctionCall {
+using namespace ABIAnalyses;
+using LatticeElement = MFI::LatticeElement;
+
+static State getMax(State Lh, State Rh) {
+  if (Lh == NoOrDead) {
+    return Rh;
+  } else if (Lh == Maybe) {
+    if (Rh == NoOrDead) {
+      return Lh;
+    } else {
+      return Rh;
+    }
+  } else {
+    return Lh;
+  }
+}
+
+LatticeElement MFI::combineValues(const LatticeElement &Lh,
+                                  const LatticeElement &Rh) const {
+
+  LatticeElement New = Lh;
+  for (const auto &KV : Rh) {
+    if (New.find(KV.first) != New.end()) {
+      New[KV.first] = getMax(New[KV.first], KV.second);
+    } else {
+      New[KV.first] = KV.second;
+    }
+  }
+  New.erase(InitialRegisterState);
+  return New;
+}
+
+bool MFI::isLessOrEqual(const LatticeElement &Lh,
+                        const LatticeElement &Rh) const {
+  if (Lh.size() > Rh.size())
+    return false;
+  for (auto &E : Lh) {
+    if (E.first == InitialRegisterState) {
+      continue;
+    }
+    if (Rh.count(E.first) == 0) {
+      return false;
+    }
+    auto RhState = Rh.lookup(E.first);
+    if (getMax(E.second, RhState) != RhState) {
+      return false;
+    }
+  }
+  return true;
+}
```

Comment:
This is quite involved code. Autogenerating this as we where doing still seems like a good idea.
One of the grand-goals of StackAnalysis2 (and ABIAnalysis2) is to be able to easily play around with analyses.
Writing all these analyses by hand is effectively making the code base more rigid than it was before.

Reply:
not sure if autogenerating code is the solution as it can easily get out of hand (e.g. a generator full of if conditions for each variation of a snippet of code, generated code that is not understandable)

I can refactor out the code that does the key by key comparison in the Maps and that should improve the readability

Reply:
Have you seen how we're doing it now? In `$BUILD_DIR/lib/StackAnalysis/ABIDataFlows.h`.
It's highly readable IMO, despite verbose. Also, generating code is less prone to copy and paste errors/forgetting edge cases.

`scripts/monotone-framework.py` doesn't handle that many special cases. If we have the need for special cases (the `-1` thing?), we should redesign the things in order not to have them.

I'd just like to autogenerate transfer functions and combine (as we're doing right now), nothing more.

Reply:
if we limit ourselves to simple lattices I agree.

I am in favour of generating the various components of the underlying lattices (the lattice when you consider a single register), just not for the product lattice

Reply:
which I think is what is happening right now

Reply:
if we limit ourselves to simple lattices I agree.

I am in favour of generating the various components of the underlying lattices (the lattice when you consider a single register), just not for the product lattice

which I think is what we are doing with `scripts/monotone-framework.py`

Reply:
Yup, exactly! Generating the product lattice is not even possible since it depends on run-time stuff, I think.

Reply:
sgtm